R/utilities.r

```diff
@@ -388,3 +400,179 @@ parse_safe <- function(text) {
   }
   out
 }
+
+switch_orientation <- function(aesthetics) {
+  # We should have these as globals somewhere
+  x <- ggplot_global$x_aes
+  y <- ggplot_global$y_aes
+  x_aes <- match(aesthetics, x)
+  x_aes_pos <- which(!is.na(x_aes))
+  y_aes <- match(aesthetics, y)
+  y_aes_pos <- which(!is.na(y_aes))
+  if (length(x_aes_pos) > 0) {
+    aesthetics[x_aes_pos] <- y[x_aes[x_aes_pos]]
+  }
+  if (length(y_aes_pos) > 0) {
+    aesthetics[y_aes_pos] <- x[y_aes[y_aes_pos]]
+  }
+  aesthetics
+}
+
+#' Utilities for working with omnidirecitonal layers
+#'
+#' These functions are what underpins the ability of certain geoms to work
+#' automatically in both directions. See the *Extending ggplot2* for how they
+#' are used when implementing `Geom`, `Stat`, and `Position` classes.
+#'
+#' `has_flipped_aes()` is used to sniff out the orientation of the layer from
+#' the data. It has a range of arguments that can be used to finetune the
+#' sniffing based on what the data should look like. `flip_data()` will switch
+#' the column names of the data so that it looks like x-oriented data.
+#' `flipped_names()` provides a named list of aesthetic names that corresponds
+#' to the orientation of the layer.
+#'
+#' @param data The layer data
+#' @param params The parameters of the `Stat`/`Geom`. Only the `orientation`
```

Comment:
Why take a list then?

Second and subsequent lines should be indented by two spaces

Reply:
The main motivation is dual... It makes the call more concise, and it allows adding further controls without rewriting all geom/stat code